File: Behavior/Memento/Lua/memento.lua
-- Memento类
local Memento = {}
Memento.__index = Memento

function Memento:new(state)
    local m = setmetatable({}, Memento)
    m.state = state
    return m
end

function Memento:getState()
    return self.state
end

-- Originator类
local Originator = {}
Originator.__index = Originator

function Originator:new()
    return setmetatable({ state = "" }, Originator)
end

function Originator:setState(state)
    self.state = state
end

function Originator:getState()
    return self.state
end

function Originator:saveToMemento()
    return Memento:new(self.state)
end

function Originator:restoreFromMemento(memento)
    self.state = memento:getState()
end

-- Caretaker类
local Caretaker = {}
Caretaker.__index = Caretaker

function Caretaker:new()
    return setmetatable({ mementos = {} }, Caretaker)
end

function Caretaker:addMemento(memento)
    table.insert(self.mementos, memento)
end

function Caretaker:getMemento(index)
    return self.mementos[index]
end

-- 使用备忘录模式的示例
local originator = Originator:new()
local caretaker = Caretaker:new()

-- 设置初始状态
originator:setState("State #1")
-- 保存状态
caretaker:addMemento(originator:saveToMemento())

-- 改变状态
originator:setState("State #2")

-- 从备忘录中恢复状态
originator:restoreFromMemento(caretaker:getMemento(1))
print("Current State: " .. originator:getState())

-- 结果输出将会是 "State #1"

File: Behavior/Memento/TypeScript/memento.ts
class Memento {
    private state: string;

    constructor(state: string) {
        this.state = state;
    }

    public getState(): string {
        return this.state;
    }
}

class Originator {
    private state: string = ''; // 提供一个默认值

    public setState(state: string): void {
        this.state = state;
        console.log(`State is set to: ${this.state}`);
    }

    public getState(): string {
        return this.state;
    }

    public save(): Memento {
        console.log('Saving state.');
        return new Memento(this.state);
    }

    public restore(memento: Memento): void {
        this.state = memento.getState();
        console.log(`State after restoring: ${this.state}`);
    }
}

class Caretaker {
    private mementos: Memento[] = [];

    public addMemento(memento: Memento): void {
        this.mementos.push(memento);
    }

    public getMemento(index: number): Memento {
        return this.mementos[index];
    }
}

// Usage
const originator = new Originator();
const caretaker = new Caretaker();

originator.setState('State #1');
caretaker.addMemento(originator.save());

originator.setState('State #2');
caretaker.addMemento(originator.save());

originator.setState('State #3');
console.log(`Current State: ${originator.getState()}`);

// Restore to State #2
originator.restore(caretaker.getMemento(1));
console.log(`Current State: ${originator.getState()}`);

// Restore to State #1
originator.restore(caretaker.getMemento(0));
console.log(`Current State: ${originator.getState()}`);
